kloppy/infra/serializers/tracking/signality.py

```diff
@@ -0,0 +1,267 @@
+        ball_position = frame["ball"]["position"]
+        if ball_position:
+            ball_coordinates = Point3D(
+                x=ball_position[0], y=ball_position[1], z=ball_position[2]
+            )
+        else:
+            ball_coordinates = None
+
+        players_data = {}
+        for ix, side in enumerate(["home", "away"]):
+            for raw_player_positional_info in frame[f"{side}_team"]:
+                player = next(
+                    (
+                        player
+                        for player in teams[ix].players
+                        if player.jersey_no
+                        == raw_player_positional_info["jersey_number"]
+                    ),
+                    None,
+                )
+                if player:
+                    player_position = raw_player_positional_info["position"]
+                    player_coordinates = Point(
+                        x=player_position[0], y=player_position[1]
+                    )
+                    player_speed = raw_player_positional_info["speed"]
+                    players_data[player] = PlayerData(
+                        coordinates=player_coordinates, speed=player_speed
+                    )
+                else:
+                    logger.debug(
+                        f"Player with jersey no: {raw_player_positional_info['jersey_number']} not found in {side} team"
+                    )
+
+        return Frame(
+            frame_id=frame_id,
+            timestamp=frame_timestamp,
+            ball_coordinates=ball_coordinates,
+            players_data=players_data,
+            period=period,
+            ball_state=None,
+            ball_owning_team=None,
+            other_data={},
+        )
+
+    @classmethod
+    def _get_signality_attacking_directions(
+        cls, frames: List[Frame], periods: List[Period]
+    ) -> Dict[int, AttackingDirection]:
+        """
+        with only partial tracking data we cannot rely on a single frame to
+        infer the attacking directions as a simple average of only some players
+        x-coords might not reflect the attacking direction.
+        """
+        attacking_directions = {}
+        frame_period_ids = np.array([_frame.period.id for _frame in frames])
+        frame_attacking_directions = np.array(
+            [
+                attacking_direction_from_frame(frame)
+                if len(frame.players_data) > 0
+                else AttackingDirection.NOT_SET
+                for frame in frames
+            ]
+        )
+
+        for period in periods:
+            period_id = period.id
+            if period_id in frame_period_ids:
+                count = Counter(
+                    frame_attacking_directions[frame_period_ids == period_id]
+                )
+                attacking_directions[period_id] = count.most_common()[0][0]
+            else:
+                attacking_directions[period_id] = AttackingDirection.NOT_SET
+
+        return attacking_directions
```

Comment:
This code seems to be copy-pasted from the SkillCorner deserializer. Can you move this to kloppy/domain/services/__init__.py?